### ISSUE TITLE
workflow UX improvements

### ISSUE THREAD

**Issue by Panaetius** [SwissDataScienceCenter/renku-python#1875]
As a result of multiple discussions over the past year, we've come to the conclusion that we don't just want to overhaul the metadata for workflows and how it's stored, but also the UX regarding workflows.

This epic serves as the central place for keeping discussions and to outline the future work to be done.

Overall Considerations/Mission Statement
================================

So far, renku hast been mostly restricted to provenance tracking of executions, which is supported well by `renku run` (and `renku rerun`/`renku update`). We have dabbled in workflow execution/workflow management a bit, with rerun/update allowing re-execution of workflows in a very limited fashion. 

This has opened up many requests for more features in this direction, like changing parameters for re-execution, importing/exporting workflows between projects, having recurrent workflows with changing output paths (e.g. daily data fetch tasks), running workflows in parallel, re-executing workflows on remote HPC clusters, to just name a few.

Renku's approach to dealing with workflows differs from regular workflow systems in that we take a more simplistic view of workflows, but we allows users to implement workflows by trying to work as normal and recording their steps as they go, as opposed to a declarative workflow language where users have to write workflow files which they can then execute. We also differ from most workflow management software in that provenance/reproducibility is at the core of our approach and not just something tacked after the fact. Additionally we (could) support provenance and workflows spanning the larger renkulab ecosystem in the form of multi-project workflows and multi-project provenance, which traditional workflow systems don't support (they're rarely tied to a platform and any glue has to be added manually by a user).

I think we should add more workflow language level features following our approach, rather than producing yet another workflow language and that these differences to regular workflow languages do set us apart and bring something valuable to the table that isn't covered by another tool already.

However, we should keep things as simple as possible. This to me means that 
- at no point should a user ever have to manually touch one of our workflow files
- we do not support complicated branching/conditional logic, nor do we support any recursive functionality
- we don't want to end up with monstrous commands that take dozens of arguments for different scenarios, rather we'd want a few, simple commands that together enable what we want to support
- command line execution remains the basic unit of work, we don't want to drop down to the script level, i.e. one workflow step equals one command line execution of an executable or script
- all the advanced functionality is based on composing this basic unit of execution in clever ways

Potential New Commands/Functionality
==============================

Naming workflows
------------------------
Users should be able to name workflows in a sensible manner that makes it easy to address a specific workflow step without having to know UUIDs or the like. Additionally, they should be able to add a description to a workflow step that describes in a human-readable way what the step intends to do. This should be supported as part of the initial `renku run --label <...> --description <...> --` as well as a command like `renku workflow label` to do this after the fact (also allowing for renames). This should also allow naming the last executed workflow (in case a user forgot). Most of the meta workflow commands below should also support naming their resulting workflows.

renku workflow group
----------------------------

Allows grouping multiple (connected?) steps into a virtual workflow. Syntax could be `renku workflow group name1...nameN`, `renku workflow group --from inputfile --to outputfile`, `renku workflow group --from-step <starting-step> --to-step <end-step>`.

We could already support this with the `renku:Run` node's `processes` property and this is what a `renku update` already results in.

The virtual workflows should be addressable by name as if they were a single step.

One question is if we want to allow multiple levels of grouping or if we unroll groups of they're specified as part of another group?

Also, should a user be able to provide a default inputmap for the group that creates group-inputs mapped to individual step inputs?  And/or a mapping of outputs of a step to inputs of a subsequent step? We could do this for files automatically with the current logic, but replacing paths with corresponding identifiers. See  the next section for a discussion of mappings.

Input/output mappings
------------------------------
inputmapping should be a yaml-like structure that maps inputs to values, e.g.

```
'0':
  files:
    inputfile: ./path/to/file
    otherinputfile: ./other/path
  parameters:
    '0': 5.7
    lr: 0.9
    alpha: 0.1
preprocessing:
  files:
  [...]
```
Which would mean that the first step has those files and values mapped to the respective parameters. We need a good automated naming scheme to give names to the inputs of a step. In the renku library, this can easily be provided by the user. In the shell, we could guess them based on the prefix of the parameter (e.g. `--config myconfig.ini` would be name `config` in the mapping) or numbering them by occurrence. Similarly, the steps can be addressed by name of the substeps or by index of the execution order.

Output mappings of files (with the current logic for detecting output files) can be achieved in a similar way, although we might need some sort of glob support to deal with variable output filenames.

Detecting output parameters could be done similarly to how github actions does it, where the code prepends values with something like https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-output-parameter (Of course, in the renku library, this is trivial). For commands that the user cannot directly change the output of (like system commands/third party executables), it might be necessary for them to wrap the call in a bash script or similar. We probably don't want to support regex matching of command output, as that would lead to too much complexity.

All of the above might also be specified at the shell level, similar to how helm `--set` works.

renku workflow execute
-------------------------------
From https://github.com/SwissDataScienceCenter/renku/issues/1929

Allows executing a workflow template/series of workflow templates on a remote cluster

Syntax:
```
renku workflow execute --provider <provider> --config <config file> <workflow name> --inputmapping <mapping> --outputmapping <mapping>
```
Config can also be stored in renku.ini (similar to kube-context for k8s). input/output mapping should map filepaths/parameter to inputs/outputs of steps.

renku workflow loop
--------------------------

Allows simple looping of a step over a variety of inputs (think `for` loop). We probably don't want to support conditional loops (like `while`), at least not for now.

Syntax:

```
renku workflow loop <workflow name> --inputmapping <mapping>
```
Where inputmapping is a list of mapping for input parameters/files that gets iterated over. Should it be a list of whole mappings or a mapping that itself contains lists in some places, and we do a cartesian product over the lists to try all combinations? Lots to discuss here.

This could potentially create a new named workflow group? Also, it'd be nice if this could be done without actually executing the commands (similar to `renku workflow group`), so that the looping virtual workflow can be created and then executed on a remote cluser using `renku workflow execute`, in cases where doing the loop is not computationally feasible locally.

Asynchronous execution
--------------------------------
Asynchronous execution of a command usually consists of 2 steps, a step that starts some job and returns an identifier for the job, and a followup step that waits for/gather the result. The first can already be done with a regular `renku run` if the job-id can be specified as an output parameter so it's available for subsequent steps.

To deal with the result of an async command, we could easily implement something that works very similar to `renku run`, but with logic for waiting for a result, e.g.

Syntax:
```
renku await --successcodes=[...] maxretries=N --timeout=300s --backoff=5s -- [...]
```

renku workflow visualize
-------------------------------
it should be possible to visualize workflows/workflow groups as a DAG to see what is happening, and to visualize individual steps in a way that shows the input and outputs and other details of the individual step.

This could just be `renku workflow visualize <name>`

The output could be something like https://stackoverflow.com/questions/5408675/ascii-visualisation-of-a-graph-of-nodes-in-python/62073648#62073648 

We'd also need a command to get neighbors/parents of a step.

renku workflow history
------------------------------
It should be possible to get a list of workflows that touched a file, to see how it was affected by them. Also to see a history of executions of a workflow and what they did. This could be `renku workflow history <file>` or `renku workflow history <workflow name>`

Parallel Execution
------------------------

We have the `--isolation` flag but we need to design this better to get a handle on what parallel execution would have to entail to work properly.

renku workflow remove
------------------------------
Removes a workflow. Do we want soft-deletion to keep provenance complete, or do we also want to delete any accompanying provenance?

renku workflow export
-----------------------------

Writes workflow file(s) in a specific workflow language for use without renku (still dependent on our Dockerfile to manage the environment).

rerun
-------

`renku rerun` could keep it's existing behavior (but switching `--edit-inputs` over to the inputmapping style), and additionally also support passing the name of a worklow/workflow group instead of a filename

update
---------
Can work as before and we probably should not change it's semantics to any of the other features. The one exception to this would be that passing workflow names to limit what gets updated (also to `--from`) might make sense.

status
--------

renku status should work pretty much as before. Only executions tracked in actual provenance matter for status, virtual workflows that were defined but never executed don't count.

Metadata Structure
===============

How to represent all this in our metadata is a big topic. `renku:Run` already supports quite a few of these use-cases, though work would have to be done to deal with the parametrization. On the provenance side, a lot would stay the same as well, but we would also need to keep track of the parameters(input/output mappings) used. Looping is another special case that we currently do not support in our metadata.

Actually running all this
==================

Since we don't have our own workflow runtime, we have to make sure that we can actually execute these workflows. Most of the concepts should be general enough that they can be translated to most workflow languages, but how do we deal with cases where this is not possible?

**Comment by schymans** [comment]
Thanks a lot for this vision of workflows in renku! In the last example (Recurring data import), I wonder how the link to external data would be saved and show up in the knowledge graph. As I understand it, linking knowledge graphs from different projects, even different renku deployments, was the great aim from the start, and I wondered if this could be the place to start implementing it. Moreover, would it be possible to not just get data from an external source, but also code? For example, if I develop a script to process a certain type of data in Repo1, then use this script in Repo2, I would like to be able to include that link in the workflow, i.e. transfer any updates to the script in Repo1 to Repo2 and then process the data in Repo2 using this updated script. Is such a use scenario on the horizon at all?

**Comment by Panaetius** [comment]
If the link to the external data was a parameter in the command used to get it (e.g. a URL), it'd show up as a parameter. In the example above it'd be hidden inside the `download.sh` file and not immediately apparent (Only that script would show up in the knowledge graph).

We do envision that the changes outlined above will lead to a command like `renku workflow import <id>` that allows importing (and updating from) a workflow from a different renku project, similar to how `renku dataset import` works. But this isn't fully fleshed out yet.

**Comment by schymans** [comment]
Actually, I didn't realise that your examples don't use any of the `renku dataset` functions. If we do `renku run wget URL`, where the URL would be the URL of a code file in a different project, would this link between project show up in the KG? Maybe I was too much fixated on the dataset method when thinking about getting code from another project.

**Comment by Panaetius** [comment]
It wouldn't exactly link to the other project, it would just show up as a string parameter value to a run in the KG
